### ISSUE TITLE
[C++] Cross compiling in Yocto is not working

### ISSUE THREAD

**Issue by pulkomandy** [apache/arrow#34093]
### Describe the bug, including details regarding any error messages, version, and platform.

Hello,

We are using Arrow on an embedded system and packaging it using Yocto. We are currently using version 9 (which works fine) and are trying to update to version 11. Unfortunately, it seems cross compilation is not working correctly.

When using "system" Thrift:

```
| CMake Error at /builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/recipe-sysroot/usr/lib/cmake/thrift/ThriftConfig.cmake:42 (message):
|   File or directory /usr/lib/cmake/thrift referenced by variable
|   THRIFT_CMAKE_DIR does not exist !
| Call Stack (most recent call first):
|   /builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/recipe-sysroot/usr/lib/cmake/thrift/ThriftConfig.cmake:59 (set_and_check)
|   cmake_modules/FindThriftAlt.cmake:56 (find_package)
|   cmake_modules/ThirdpartyToolchain.cmake:261 (find_package)
|   cmake_modules/ThirdpartyToolchain.cmake:1538 (resolve_dependency)
|   CMakeLists.txt:498 (include)
```

Checking the value of THRIFT_CMAKE_DIR, we see that it is set to /usr/lib/cmake/thrift, but seems to ignore the Yocto sysroot.

We have tried using the bundled Thrift as well, and we get a different error:

```
| FAILED: lib/libthrift.a
| : && /builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/recipe-sysroot-native/usr/bin/cmake -E rm -f lib/libthrift.a && /builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/build/thrift_ep-prefix/src/thrift_ep-build/aarch64-fsl-linux-gcc-ar qc lib/libthrift.a  lib/cpp/CMakeFiles/thrift.dir/src/thrift/TApplicationException.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/TOutput.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/async/TAsyncChannel.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/async/TAsyncProtocolProcessor.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/async/TConcurrentClientSyncInfo.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/concurrency/ThreadManager.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/concurrency/TimerManager.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/processor/PeekProcessor.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/protocol/TBase64Utils.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/protocol/TDebugProtocol.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/protocol/TJSONProtocol.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/protocol/TMultiplexedProtocol.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/protocol/TProtocol.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TTransportException.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TFDTransport.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TSimpleFileTransport.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/THttpTransport.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/THttpClient.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/THttpServer.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TSocket.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TSocketPool.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TServerSocket.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TTransportUtils.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TBufferTransports.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/SocketCommon.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/server/TConnectedClient.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/server/TServerFramework.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/server/TSimpleServer.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/server/TThreadPoolServer.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/server/TThreadedServer.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TPipe.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TPipeServer.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TFileTransport.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/VirtualProfiling.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/server/TServer.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TSSLSocket.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TSSLServerSocket.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/transport/TWebSocketServer.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/concurrency/ThreadFactory.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/concurrency/Thread.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/concurrency/Monitor.cpp.o lib/cpp/CMakeFiles/thrift.dir/src/thrift/concurrency/Mutex.cpp.o && /builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/recipe-sysroot-native/usr/bin/aarch64-fsl-linux/aarch64-fsl-linux-ranlib lib/libthrift.a && :
| /bin/sh: 1: /builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/build/thrift_ep-prefix/src/thrift_ep-build/aarch64-fsl-linux-gcc-ar: not found
| ninja: build stopped: subcommand failed.
| 
| -- stderr output is:
| 
| CMake Error at /builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/build/thrift_ep-prefix/src/thrift_ep-stamp/thrift_ep-build-RELEASE.cmake:47 (message):
|   Stopping after outputting logs.
| 
| 
| ninja: build stopped: subcommand failed.
| WARNING: exit code 1 from a shell command.
| 
```

For some reason, the build attempts to use the command `aarch64-fsl-linux-gcc-ar` with a prefix pointing to the current directory, instead of the expected toolchain directory.

We are trying to investigate this, can you provide some help or recommendations on cross compiling Arrow?

### Component(s)

C++

**Comment by assignUser** [comment]
Without looking into this in detail (I'll come back to this later) we switched from C++11 to C++17 for 10.0.0 as this is a major change I wanted to mention it :)

**Comment by kou** [comment]
> When using "system" Thrift:

How did you install Thrift? Did you install Thrift with `DESTDIR`?

> We have tried using the bundled Thrift as well, and we get a different error:

Could you show `/builds/liebherr-mining/connectivitybox/build/tmp_s32g274aconbox2_liebherr/work/cortexa53-crypto-fsl-linux/apache-arrow/11.0.0-r0/build/thrift_ep-prefix/src/thrift_ep-stamp/thrift_ep-configure-RELEASE.cmake`?

**Comment by pulkomandy** [comment]
I investigated a bit further and it turns out to be a bug in Yocto, which did not set the CMAKE_AR variable correctly. It was fixed by this change which I backported into our version of Yocto: https://git.yoctoproject.org/poky/commit/meta/classes/cmake.bbclass?h=hardknott&id=887e56c1b274db773ad8f9925e222a06992cf650

Nothing is wrong on Arrow and Thrift side for the bundled case which is now working fine. I did not investigate the problem with using the "system" Thrift (provided by Yocto) but there isn't any reason to use it in our case, so I will go with the Bundled version and close this issue.

Thanks for your help and sorry for reporting this in the wrong place since it wasn't your fault.
